packages/core/src/plugins/normalize/withStagingEditor.ts

```diff
@@ -0,0 +1,175 @@
+  //     //   return !!op.normalizingPluginKeys[p.key];
+  //     // });
+  //
+  //     const [node, path] = entry;
+  //
+  //     const id = node.id ?? nanoid();
+  //
+  //     if (!node.id) {
+  //       setNodes<TNode>(
+  //         editor,
+  //         {
+  //           id,
+  //         },
+  //         { at: path }
+  //       );
+  //     }
+  //
+  //     if (node.id) {
+  //       const iterationCountByNodeId =
+  //         iterationCountByNodeIdByPlugin.get(plugin) ?? {};
+  //
+  //       const iterationCount = iterationCountByNodeId[node.id] ?? 0;
+  //     }
+  //
+  //     if (maxIterations && normalizingPluginOperations.length > maxIterations) {
+  //       console.warn(
+  //         `Plugin ${plugin.key} could not completely normalize the node after ${maxIterations} iterations! This is usually due to incorrect normalization logic that leaves a node in an invalid state.`
+  //       );
+  //       return false;
+  //     }
+  //
+  //     if (query && !query(entry as any)) {
+  //       return false;
+  //     }
+  //
+  //     editor.normalizingPluginKeys[p.key] = true;
+  //
+  //     const stop = (value: boolean) => {
+  //       delete editor.normalizingPluginKeys[p.key];
+  //       return value;
+  //     };
+  //
+  //     if (applyNormalizeNode(entry as any)) {
+  //       plugin.normalizingCount = (plugin.normalizingCount ?? 0) + 1;
+  //       // iterationCountByNodeIdByPlugin.set(plugin, {
+  //       //   ...iterationCountByNodeId,
+  //       //   [key]: iterationCount + 1,
+  //       // });
+  //       return stop(true);
+  //     }
+  //
+  //     return stop(false);
+  //   });
+  //
+  //   if (normalized) return;
+  //
+  //   normalizeNode(entry);
+  // };
+
+  const stagingEditorPlugins = editor.plugins.filter(
+    (p) => p.key !== KEY_STAGING
+  );
+
+  editor.errors = [];
+
+  editor.stagingEditor = createPlateEditor({
+    id: 'staging',
+    plugins: stagingEditorPlugins,
+    disableCorePlugins: {
+      staging: true,
+    },
+  });
+
+  const resetStagingEditor = () => {
+    const { id, key: _key, ...cloneEditor } = cloneDeep(editor);
+    Object.keys(cloneEditor).forEach((key) => {
+      if (typeof cloneEditor[key] === 'function') delete cloneEditor[key];
+    });
+
+    Object.assign(editor.stagingEditor, {
+      ...editor.stagingEditor,
+      ...cloneEditor,
+    });
+  };
+
+  editor.setNormalizing = (value) => {
+    editor.stagingEditor.setNormalizing(value);
+    setNormalizing(value);
+  };
+
+  editor.withoutNormalizing = (fn) => {
+    const value = editor.isNormalizing();
+    editor.setNormalizing(false);
+    try {
+      fn();
+    } finally {
+      editor.setNormalizing(value);
+    }
+
+    let hasError = false;
+
+    try {
+      editor.stagingEditor.normalize();
+    } catch (err) {
+      editor.errors.push({ type: 'normalize', error: err });
+      hasError = true;
+    }
+
+    if (!hasError) {
+      editor.normalize();
+    }
+  };
```

Comment:
@12joan here is my first try to clone `withoutNormalizing`, but it looks like I'm missing something, looking at the failing test. Would it be related to https://github.com/ianstormtaylor/slate/blob/3243c7e34ac2602618c67c88b1b7df07fde1c2ec/packages/slate/src/editor/normalize.ts#L13 ?

Reply:
It looks like `withoutNormalizing` is being called by both `insertFragment` and `insertNodes`. Is it possible the nested calls are causing problems with the staging editor?

<img width="411" alt="image" src="https://user-images.githubusercontent.com/4272090/234550666-f8b97e9c-f67d-4881-8894-21dcd17d7e0f.png">

Reply:
Here's the command I'm using, if you want to try it yourself. Your IDE probably has a built-in debugger client, but I'm using `chrome:inspect`.

```bash
yarn node --inspect-brk ./node_modules/.bin/jest --runInBand packages/core/src/plugins/normalize/withStagingEditor.spec.tsx -t 'insertFragment'
```

Reply:
In the following `editor.apply` call (which is the one that errors):

- spec
  - insertFragment
    - withoutNormalizing
      - normalize
        - withoutNormalizing
          - normalizeNode
            - mergeNodes
              - withoutNormalizing
                - apply

`editor.children` has '', 'a', 'b', but `editor.stagingEditor.children` has 'ab'. Shouldn't they be the same at the start of `apply`?

Reply:
Looks like it has already been merged (normalized) on the staging editor despite using `setNormalize(false)`

Reply:
Ah! We probably need to override `editor.stagingEditor.withoutNormalizing`